libcrux-ml-kem/src/polynomial.rs

```diff
@@ -147,107 +155,338 @@ pub(crate) const fn vec_len_bytes<const K: usize, Vector: Operations>() -> usize
 /// Given two polynomial ring elements `lhs` and `rhs`, compute the pointwise
 /// sum of their constituent coefficients.
 #[inline(always)]
-#[hax_lib::fstar::verification_status(lax)]
+#[hax_lib::fstar::options("--z3rlimit 300")]
+#[hax_lib::requires(fstar!(r#"
+        forall (i:nat). i < v ${VECTORS_IN_RING_ELEMENT} ==>
+         (let lhs_i = i1.f_to_i16_array (${myself}.f_coefficients.[ sz i ]) in
+          let rhs_i = i1.f_to_i16_array (${rhs}.f_coefficients.[ sz i ]) in
+          Libcrux_ml_kem.Vector.Traits.Spec.add_pre lhs_i rhs_i)"#))]
+#[hax_lib::ensures(|_| fstar!(r#"
+        forall (i:nat). i < v ${VECTORS_IN_RING_ELEMENT} ==>
+         (let lhs_i = i1.f_to_i16_array (${myself}.f_coefficients.[ sz i ]) in
+          let rhs_i = i1.f_to_i16_array (${rhs}.f_coefficients.[ sz i ]) in
+          let result_i = i1.f_to_i16_array (${myself}_future.f_coefficients.[ sz i ]) in
+          Libcrux_ml_kem.Vector.Traits.Spec.add_post lhs_i rhs_i result_i)"#))]
 fn add_to_ring_element<Vector: Operations, const K: usize>(
     myself: &mut PolynomialRingElement<Vector>,
     rhs: &PolynomialRingElement<Vector>,
 ) {
+    #[cfg(hax)]
+    let _myself = myself.coefficients;
     for i in 0..myself.coefficients.len() {
+        hax_lib::loop_invariant!(|i: usize| {
+            fstar!(
+                r#"
+                v $i <= v ${VECTORS_IN_RING_ELEMENT} /\
+                (forall (j:nat). (j >= v $i /\ j < v ${VECTORS_IN_RING_ELEMENT}) ==>
+                   (let _myself_j = i1.f_to_i16_array (${_myself}.[ sz j ]) in
+                    let myself_j = i1.f_to_i16_array (${myself}.f_coefficients.[ sz j ]) in
+                    let rhs_j = i1.f_to_i16_array (${rhs}.f_coefficients.[ sz j ]) in
+                    myself_j ==  _myself_j /\
+                    Libcrux_ml_kem.Vector.Traits.Spec.add_pre myself_j rhs_j)) /\
+                (forall (j:nat). j < v $i ==>
+                   (let _myself_j = i1.f_to_i16_array (${_myself}.[ sz j ]) in
+                    let myself_j = i1.f_to_i16_array (${myself}.f_coefficients.[ sz j ]) in
+                    let rhs_j = i1.f_to_i16_array (${rhs}.f_coefficients.[ sz j ]) in
+                    Libcrux_ml_kem.Vector.Traits.Spec.add_post _myself_j rhs_j myself_j))"#
+            )
+        });
+
         myself.coefficients[i] = Vector::add(myself.coefficients[i], &rhs.coefficients[i]);
     }
 }
 
 #[inline(always)]
-#[hax_lib::fstar::verification_status(lax)]
+#[hax_lib::requires(fstar!(r#"is_bounded_poly 28296 ${myself}"#))]
+#[hax_lib::ensures(|_| fstar!(r#"is_bounded_poly 3328 ${myself}_future"#))]
 fn poly_barrett_reduce<Vector: Operations>(myself: &mut PolynomialRingElement<Vector>) {
-    // Using `hax_lib::fstar::verification_status(lax)` works but produces an error while extracting
+    let _myself = myself.coefficients;
```

Comment:
Add a `cfg(hax)` for these unused variables?

Reply:
Done.